fix($parse): always re-evaluate filters within literals when an input is an object

jbedard opened the pull request:
Fixes #15964 

This changes the [compareObjectIdentity](https://github.com/angular/angular.js/blob/v1.6.4/src/ng/parse.js#L1785) flag from being per-expression to per-input.  This allows inputs to filters to not compare object entity, but other inputs such as `- x` or `{x: x}` to only compare object entity.  This also expands this `compareObjectIdentity = true` beyond just object literals.

jbedard committed:
fix($parse): always re-evaluate filters within literals when an input is an object

Fixes #15964
Closes #15990

jbedard commented on src/ng/parse.js:
```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isShallowState(stack) {
```
This could probably use a better name/terminology...

petebacondarwin replied:
`isStackPure`?

jbedard commented on src/ng/parse.js:
```diff
@@ -781,7 +819,7 @@ ASTCompiler.prototype = {
       fn: {vars: [], body: [], own: {}},
       assign: {vars: [], body: [], own: {}},
-      inputs: []
+      inputs: {}
```
Probably a nicer way of doing this...

lgalfaso replied:
By doing this (and later `Object.keys`) you are abusing the fact that `Object.keys` returns the keys in the order that were added to assert that expressions as `foo.bar();foo.baz()` will be evaluated in the right order. Even when this is somehow true, this is not part of the spec and we should not depend on this out-of-spec fact.

jbedard replied:
I forgot the order even mattered, good point 👍

petebacondarwin commented on test/ng/parseSpec.js:
```diff
@@ -2973,6 +2973,111 @@ describe('parser', function() {
+            ' when the instance changes but valueOf() does not', inject(function($parse) {
+          var filterCalls = 0;
+          $filterProvider.register('foo', valueFn(function(input) {
+            filterCalls++;
```
`filterCalls` is never tested, is it needed?

jbedard commented on test/ng/parseSpec.js:
```diff
@@ -2973,6 +2973,111 @@ describe('parser', function() {
+          expect(watcherCalls).toBe(2);
+        }));
+
+        it('should always reevaluate filters with literal input containing non-primitives',
```
One way to test this better (which I'll add) would be something like `[(obj | filterThatConvertsToPrimitive)]`

petebacondarwin commented on src/ng/parse.js:
```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isShallowState(stack) {
+  for (var i = 0; i < stack.length; i++) {
```
Do we need to walk the stack every time? It seems that we are either pushing or popping items on and off this stack and then re-evaluating the shallowStateness of the stack. Can we not cache the state of each stack and then we only need to evaluate the pushed item?

jbedard replied:
Yeah that's probably true.  Probably store the result of the switch statement in the stack? I'll try it out...

lgalfaso commented on src/ng/parse.js:
```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
+  return true;
+}
+
+function findConstantAndWatchExpressions(ast, $filter, stack) {
```
I am not following completely the logic here, but by just looking at the code it looks like `stack` can be replace with a simple boolean (that the first called call with `true` and the rest do `ast.isShallowState = parentShallowState && isShallowState([ast]);` (and of course this can be simplified).

jbedard replied:
Yeah I think that's possible. I guess the idea of the `stack` was that it could also potentially be used for other purposes, but I'll try your suggestion.

petebacondarwin replied:
Yeah, I think this is what I was getting at with https://github.com/angular/angular.js/pull/15990#discussion_r116576729

jbedard commented:
I think I've addressed all the comments. Also added a `ng-class` test for the exact issue someone reported.

I'm also open to other ideas around this and similar issues such as #15905 if anyone has alternative ideas.

Note that if the $filter solution in this PR is extended to interceptors it will also fix #15905 *except* [one-time literals containing objects](https://github.com/angular/angular.js/blob/v1.6.4/test/ng/directive/ngClassSpec.js#L612). That one failure is caused by [this condition](https://github.com/angular/angular.js/blob/v1.6.4/src/ng/parse.js#L1913-L1915) which I really wish I could find an alternative solution for, especially since that condition forces workarounds in multiple places where interceports are used.  The other thing that would fix that one ng-class test is [dropping nested object support in ng-class](https://github.com/angular/angular.js/issues/15960) 😒

petebacondarwin commented on src/ng/parse.js:
```diff
@@ -622,15 +622,43 @@ function isStateless($filter, filterName) {
   var argsToWatch;
   var isStatelessFilter;
+
+  ast.isPure = parentIsPure = isPure(ast, (undefined === parentIsPure) || parentIsPure);
```
`(undefined === parentIsPure) || parentIsPure` -  I would move this logic into the `isPure()` function.

petebacondarwin commented on src/ng/parse.js:
```diff
@@ -622,15 +622,43 @@ function isStateless($filter, filterName) {
+      break;
+
+    case AST.UnaryExpression:
+      return true;
```
In this case would you not defer to the `parentPureness` also?

jbedard replied:
The idea is to allow shallow-watching the `input` in things like `fn(!input)` where `!parentIsPure`, but the child is.

I've tried to exclude case which can invoke `toString` like `"" + input` because those could be non-pure (when `toString` reads inner state).

petebacondarwin commented on src/ng/parse.js:
```diff
@@ -839,13 +867,16 @@ ASTCompiler.prototype = {
-    forEach(fns, function(name) {
-      result.push('var ' + name + '=' + self.generateFunction(name, 's'));
+    forEach(inputs, function(input) {
+      result.push('var ' + input.name + '=' + self.generateFunction(input.name, 's'));
```
is there any performance/code-size gain in setting and using `var name = input.name;` here?
Or does the minification process deal with that automatically?

jbedard replied:
I think performance/code-size has no (noticeable) gain or loss either way, so I'd vote for whatever is more readable/prettier. My preference is how it is (no local var) but I can change it if others would prefer that.

petebacondarwin commented on test/ng/parseSpec.js:
```diff
@@ -2973,6 +2973,172 @@ describe('parser', function() {
           expect(watcherCalls).toBe(1);
         }));
 
+        it('should not reevaluate filters in literals with non-primitive input that does support valueOf()',
```
can we use some describe blocks to simplify the test descriptions here:

```
describe('containing literals', function() {
  describe('with non-primitive inputs', function() {
    describe('that support valueOf()', function() {
      it('should not reevaluate filters, inputs do not change', ...);
      it('should reevaluate filters, when an input valueOf() result changes', ...);
      it('should not reevaluate filters, when the input changes but valueOf() does not ', ...);
    });
  });
  describe('with non-primitive inputs', function() {
  });
});
```

you get the idea...

petebacondarwin reviewed (commented):
This looks pretty good; a few minor comments.
My main question is about the purity of `UnaryExpressions`.
Otherwise happy to land this.

jbedard committed:
test($parse): reorganize $parse tests

jbedard commented:
Updated based on PR feedback. Also added more tests and added a separate commit to reorganize the tests. How those tests are organized is debatable but I think what I did is similar to the suggestion in the comment.

petebacondarwin reviewed (approved):
LGTM!
Thanks @jbedard - please squash and merge (assuming Travis is happy).

jbedard commented:
@lgalfaso have a chance to look at the latest update?

lgalfaso commented on src/ng/parse.js:
```diff
@@ -622,15 +622,44 @@ function isStateless($filter, filterName) {
   var argsToWatch;
   var isStatelessFilter;
+
+  ast.isPure = parentIsPure = isPure(ast, parentIsPure);
```
[nit] I find it slightly odd to change the value of the argument. I would just use `ast.isPure` below and keep the argument as is.

jbedard replied:
I think I agree with that.  Could also change the `parentIsPure` param to `parentAst`, but I'll go with your suggestion unless someone says otherwise.

lgalfaso reviewed (approved):
LGTM!

Just one minor comment.

jbedard commented:
Made the change and merged into master and v1.6.x

gkalpak commented on src/ng/parse.js:
```diff
@@ -622,15 +622,44 @@ function isStateless($filter, filterName) {
+
+    // Unary always convert to primative
+    case AST.UnaryExpression:
+      return true;
```
Can't unary expressions invoke a stateful `valueOf()`?

jbedard replied:
Yeah I think you're right.

Seems like any numeric operator (everything except `!`, `&&` and `||`?) will use `valueOf()`. However this [will be handled](https://github.com/angular/angular.js/blob/b5118ac6a9e0a327b31094b3fdcdc0432b23ad2f/src/ng/parse.js#L1829) so I don't think it will lead to bugs, but this method is still wrong.

Now that I type that... `&&` and `||` also don't convert to primitives and might return an object so those are also incorrect, but [we watch the full logical expression](https://github.com/angular/angular.js/blob/b5118ac6a9e0a327b31094b3fdcdc0432b23ad2f/src/ng/parse.js#L686) so it won't lead to bugs, but this method is still wrong.

So I don't think there's a bug but it might be nice to make this function more correct?

Once again, always wait for @gkalpak to review your PRs...

gkalpak replied:
First of all, I don't know what `&&` and `||` have to do with this (since they are not unary operators) :confused:

Regarding `getValueOf()`, I don't think it is enough to "save" us (in cases where `valueOf()` returns non-primitive values). I admit it is very unlikely that this will affect any real app, but from a theoretical point of view, I think expressions such as `'+obj'`, where `obj` has a `valueOf()` method that returns an object will not be handled correctly (e.g. if the returned object stays the same by reference, but one of its properties changes).

gkalpak replied:
Hm...OK maybe that is not true, because unary operators will always evaluate to the same thing for objects whose `valueOf` returns a non-primitve value (`!` will return `false`, `+/-` will return `NaN`), so this might actually work correctly :grin:

gkalpak replied:
> I do still think that numeric unary/binary operators are non-pure because valueOf might be stateful. Which is a bug in isPure, but the use of getValueOf later on prevents a bug in $parse. I'll probably still correct the isPure function...

That's also my conclusion :smiley:
